flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptivebatch/DefaultVertexParallelismDecider.java

```diff
@@ -0,0 +1,132 @@
+import org.apache.flink.configuration.JobManagerOptions;
+import org.apache.flink.configuration.MemorySize;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+
+/** Default implementation of {@link VertexParallelismDecider}. */
+public class DefaultVertexParallelismDecider implements VertexParallelismDecider {
+
+    private static final Logger LOG =
+            LoggerFactory.getLogger(DefaultVertexParallelismDecider.class);
+
+    private final int maxParallelism;
+    private final int minParallelism;
+    private final long dataVolumePerTask;
+    private final int defaultSourceParallelism;
+
+    public DefaultVertexParallelismDecider(
+            int maxParallelism,
+            int minParallelism,
+            MemorySize dataVolumePerTask,
+            int defaultSourceParallelism) {
+
+        this.maxParallelism = maxParallelism;
+        this.minParallelism = minParallelism;
+        this.dataVolumePerTask = dataVolumePerTask.getBytes();
+        this.defaultSourceParallelism = defaultSourceParallelism;
+    }
+
+    @Override
+    public int decideParallelismForVertex(List<BlockingResultInfo> consumedResults) {
+
+        if (consumedResults.isEmpty()) {
+            // source job vertex
+            return defaultSourceParallelism;
+        } else {
+            return calculateParallelism(consumedResults);
+        }
+    }
+
+    private int calculateParallelism(List<BlockingResultInfo> consumedResults) {
+
+        long broadcastBytes =
+                consumedResults.stream()
+                        .filter(BlockingResultInfo::isBroadcast)
+                        .mapToLong(
+                                consumedResult ->
+                                        consumedResult.getBlockingPartitionSizes().stream()
+                                                .reduce(0L, Long::sum))
+                        .sum();
+
+        long nonBroadcastBytes =
+                consumedResults.stream()
+                        .filter(consumedResult -> !consumedResult.isBroadcast())
+                        .mapToLong(
+                                consumedResult ->
+                                        consumedResult.getBlockingPartitionSizes().stream()
+                                                .reduce(0L, Long::sum))
+                        .sum();
+
+        if (broadcastBytes > dataVolumePerTask
+                || (broadcastBytes == dataVolumePerTask && nonBroadcastBytes > 0)) {
+            LOG.warn(
+                    "The minimum size of one task to process is larger than "
+                            + "the size of data volume which is configured by "
+                            + "'"
+                            + JobManagerOptions.ADAPTIVE_BATCH_SCHEDULER_DATA_VOLUME_PER_TASK.key()
+                            + "'. "
+                            + "Parallelism will be set to {}.",
+                    maxParallelism);
+
+            return maxParallelism;
```

Comment:
I think it's worth discussing. When the `broadcastBytes` is close to the `dataVolumePerTask `, even if the `nonBroadcastBytes` is very small, it may cause a large parallelism, which is unnecessary and will bring great network overhead.

To solve this problem, one possible approach is to reserve a minimum ratio for the amount of non-broadcast data processed by one task. For example, it is specified that the proportion of non-broadcast data processed by one task is at least 50%, once the amount of broadcast data exceeds 50%, the parallelism should be `nonBroadcastBytes / (dataVolumePerTask * 0.5)`. In this way, when `nonBroadcastBytes` is rarely or even 0, the parallelism will be determined as `minParallelism`.

Reply:
Yes I agree that when the `broadcastBytes` is too close to `dataVolumePerTask`, it is still a problem.

In my understanding, broadcast dataset is usually relatively small against the other co-processed dataset. Therefore, I think we can introduce a cap ratio of how much the `broadcastBytes` will affect computing the parallelism. By default, the cap ratio can be `0.5` because we usually expect the broadcast dataset to be smaller. The value can be hard coded in the first version. We can make it configurable later if we see users requesting for it.